pkg/liquidity-mining/contracts/admin/AuthorizerAdaptorEntrypoint.sol

```diff
@@ -0,0 +1,106 @@
+    using Address for address;
+
+    IAuthorizerAdaptor private immutable _adaptor;
+    IVault private immutable _vault;
+
+    constructor(IAuthorizerAdaptor adaptor) {
+        _adaptor = adaptor;
+        _vault = adaptor.getVault();
+    }
+
+    /**
+     * @notice Returns the Balancer Vault
+     */
+    function getVault() public view override returns (IVault) {
+        return _vault;
+    }
+
+    /**
+     * @notice Returns the Authorizer
+     */
+    function getAuthorizer() public view override returns (IAuthorizer) {
+        return getVault().getAuthorizer();
+    }
+
+    function _canPerform(
+        bytes32 actionId,
+        address account,
+        address where
+    ) internal view returns (bool) {
+        return getAuthorizer().canPerform(actionId, account, where);
+    }
+
+    /**
+     * @notice Returns the action ID associated with calling a given function through the authorizer adaptor.
+     * @dev As the contracts managed by the adaptor don't have action ID disambiguators, we use the adaptor's globally.
+     * This means that contracts with the same function selector will have a matching action ID:
+     * if granularity is required then permissions must not be granted globally in the Authorizer.
+     *
+     * The adaptor entrypoint does not hold a disambiguator of its own; this function just forwards the call to the
+     * adaptor itself.
+     *
+     * @param selector - The 4 byte selector of the function to be called using `performAction`
+     * @return The associated action ID
+     */
+    function getActionId(bytes4 selector) public view override returns (bytes32) {
+        return _adaptor.getActionId(selector);
+    }
+
+    /**
+     * @notice Performs an arbitrary function call on a target contract, provided the caller is authorized to do so.
+     * @param target - Address of the contract to be called
+     * @param data - Calldata to be sent to the target contract. It should be at least 4 bytes long (i.e. the length of
+     * the selector corresponding to the function to be called)
+     * @return The bytes encoded return value from the performed function call
+     */
+    function performAction(address target, bytes calldata data)
+        external
+        payable
+        override
+        nonReentrant
+        returns (bytes memory)
+    {
+        // We want to check that the caller is authorized to call the function on the target rather than this function.
+        // We must then pull the function selector from `data` rather than `msg.sig`.
+        // Note that if `data` is less than 4 bytes long this will revert.
+        bytes4 selector = data[0] | (bytes4(data[1]) >> 8) | (bytes4(data[2]) >> 16) | (bytes4(data[3]) >> 24);
+
+        _require(_canPerform(getActionId(selector), msg.sender, target), Errors.SENDER_NOT_ALLOWED);
+
+        // We don't check that `target` is a contract so all calls to an EOA will succeed.
+        return target.functionCallWithValue(data, msg.value);
```

Comment:
yes, that's correct.
By doing this call properly, `canPerform` will be called twice in the authorizer, so we'll need to do the changes in the authorizer / deployment steps as well.

https://github.com/balancer-labs/balancer-v2-monorepo/pull/1977/commits/6d8f3e2805321ca0f362bb07534f7c80bd9ef178 has the basics to move forward. The current tests in the entrypoint still pass, but there's some more work to be done to adapt the existing tests to the entrypoint, since now calls that go through the adaptor directly will revert.